lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "format.h"
+
+namespace Fortran::common {
+
+template<typename CHAR> CHAR FormatValidator<CHAR>::NextChar() {
+  for (++cursor_; cursor_ < end_; ++cursor_) {
+    if (*cursor_ != ' ') {
+      return toupper(*cursor_);
+    }
+  }
+  cursor_ = end_;
+  return ' ';
+};
+
+template<typename CHAR> CHAR FormatValidator<CHAR>::LookAheadChar() {
+  for (laCursor_ = cursor_ + 1; laCursor_ < end_; ++laCursor_) {
+    if (*laCursor_ != ' ') {
+      return toupper(*laCursor_);
+    }
+  }
+  laCursor_ = end_;
+  return ' ';
+};
+
+template<typename CHAR> void FormatValidator<CHAR>::Advance(TokenKind tk) {
+  cursor_ = laCursor_;
```

Comment:
From my understanding of this logic, for every `NextChar()` call there must be a call to `LookAheadChar()` before calling `Advanced()` in order for it to work correctly. It would be better to add this information next to the definition and/or declaration of Advance(), or use an alternate logic which doesn't require this.

Reply:
I've added a comment before the Advance definition.